File: debezium/connector/kafka/DebeziumKafkaConsumer.java
package com.tiket.tix.hotel.common.debezium.connector.kafka;

import com.tiket.tix.hotel.common.debezium.DebeziumHelper;
import com.tiket.tix.hotel.common.debezium.config.DebeziumConfiguration;
import com.tiket.tix.hotel.common.debezium.connector.embedded.EmbeddedConnectorRunner;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;

import java.time.Duration;
import java.util.Collections;
import java.util.Properties;

public class DebeziumKafkaConsumer {
  private final KafkaConsumer<String, String> consumer;
  private final EmbeddedConnectorRunner runner;
  private final long durationPollMillis;

  public DebeziumKafkaConsumer(DebeziumConfiguration config, EmbeddedConnectorRunner runner) {
    this.runner = runner;

    Properties props = new Properties();
    props.put("bootstrap.servers", config.getKafkaConfiguration().getBootstrapServers());
    props.put("group.id", config.getKafkaConfiguration().getGroupId());
    props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
    props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");

    this.durationPollMillis = config.getKafkaConfiguration().getPollIntervalMs();
    this.consumer = new KafkaConsumer<>(props);
    this.consumer.subscribe(Collections.singletonList(config.getKafkaConfiguration().getTopic()));
  }

  public void startPolling() {
    new Thread(() -> {
      while (true) {
        ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(durationPollMillis));
        for (ConsumerRecord<String, String> record : records) {
          DebeziumHelper.sendMetric("kafka_consumer_received");
          try {
            runner.handleChangeFromJson(record.value());
            DebeziumHelper.sendMetric("kafka_consumer_processed");
          } catch (Exception e) {
            DebeziumHelper.sendMetric("kafka_consumer_failed");
          }
        }
      }
    }).start();
  }
}

File: debezium/connector/kafka/debezium_kafka_consumer.go
package kafka

import (
	"context"
	"time"

	"github.com/Shopify/sarama"
	"github.com/tiket/TIX-HOTEL-UTILITIES-GO/debezium/config"
	"github.com/tiket/TIX-HOTEL-UTILITIES-GO/debezium/connector/embedded"
	"github.com/tiket/TIX-HOTEL-UTILITIES-GO/debezium/helper"
)

type DebeziumKafkaConsumer struct {
	consumer   sarama.ConsumerGroup
	runner     *embedded.UpsertConnectorRunner
	topic      string
	pollMillis time.Duration
	cfg        *config.DebeziumConfig
}

func NewDebeziumKafkaConsumer(cfg *config.DebeziumConfig, runner *embedded.UpsertConnectorRunner) *DebeziumKafkaConsumer {
	consumerGroup, err := sarama.NewConsumerGroup([]string{cfg.KafkaConfig.Bootstrap}, cfg.KafkaConfig.GroupID, nil)
	if err != nil {
		panic(err)
	}
	return &DebeziumKafkaConsumer{
		consumer:   consumerGroup,
		runner:     runner,
		topic:      cfg.KafkaConfig.Topic,
		pollMillis: time.Millisecond * time.Duration(cfg.KafkaConfig.PollIntervalMs),
		cfg:        cfg,
	}
}

func (c *DebeziumKafkaConsumer) StartPolling() {
	ctx := context.Background()
	for {
		handler := &consumerGroupHandler{runner: c.runner}
		if err := c.consumer.Consume(ctx, []string{c.topic}, handler); err != nil {
			helper.SendMetric(helper.TagBuilder{
				EventType: "kafka_consumer_error",
			})
		}
		time.Sleep(c.pollMillis)
	}
}

type consumerGroupHandler struct {
	runner *embedded.UpsertConnectorRunner
}

func (h *consumerGroupHandler) Setup(sarama.ConsumerGroupSession) error   { return nil }
func (h *consumerGroupHandler) Cleanup(sarama.ConsumerGroupSession) error { return nil }
func (h *consumerGroupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
	for msg := range claim.Messages() {
		helper.SendMetric(helper.TagBuilder{
			EventType: "kafka_consumer_message_received",
		})
		h.runner.HandleChangeFromJSON(string(msg.Value))
		helper.SendMetric(helper.TagBuilder{
			EventType: "kafka_consumer_message_processed",
		})
		sess.MarkMessage(msg, "")
	}
	return nil
}
